File: src/nodejs_demo/ts_src/mdif.ts
/*******************************************************************************
 *                                                                             *
 *                                                 ,,                          *
 *                                                       ,,,,,                 *
 *                                                           ,,,,,             *
 *           ,,,,,,,,,,,,,,,,,,,,,,,,,,,,                        ,,,,          *
 *          ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,            ,,,,          ,,,,        *
 *          ,,,,,       ,,,,,      ,,,,,,                ,,,,        ,,,       *
 *          ,,,,,       ,,,,,      ,,,,,,                   ,,,        ,,,     *
 *          ,,,,,       ,,,,,      ,,,,,,       ,,,           ,,,        ,     *
 *          ,,,,,       ,,,,,      ,,,,,,           ,,,         ,,        ,    *
 *          ,,,,,       ,,,,,      ,,,,,,              ,,        ,,            *
 *          ,,,,,       ,,,,,      ,,,,,,                ,        ,            *
 *          ,,,,,       ,,,,,      ,,,,,,                 ,                    *
 *          ,,,,,       ,,,,,      ,,,,,,                                      *
 *          ,,,,,       ,,,,,      ,,,,,,                                      *
 *                                       ,,,,,,,,,,,,,,,,,,,,,,,,,,            *
 *                                       ,,,,,,,,,,,,,,,,,,,,,,,,,,,,          *
 *                                       ,,,,,                  ,,,,,,         *
 *                     ,                 ,,,,,                  ,,,,,,         *
 *             ,        ,,               ,,,,,                  ,,,,,,         *
 *    ,        ,,        ,,,             ,,,,,                  ,,,,,,         *
 *     ,        ,,,         ,,,          ,,,,,                  ,,,,,,         *
 *     ,,,       ,,,                     ,,,,,                  ,,,,,,         *
 *      ,,,        ,,,,                  ,,,,,                  ,,,,,,         *
 *        ,,,         ,,,,               ,,,,,                  ,,,,,,         *
 *         ,,,,,            ,,,,         ,,,,,,,,,,,,,,,,,,,,,,,,,,,,          *
 *            ,,,,                       ,,,,,,,,,,,,,,,,,,,,,,,,,,            *
 *               ,,,,,                                                         *
 *                    ,,,,,                                                    *
 *                                                                             *
 * Program/file : mdif.ts                                                      *
 *                                                                             *
 * Description  :  Provides MDIF interface via a socket                        *
 *              :                                                              *
 *                                                                             *
 * Copyright 2023 MyDefence A/S.                                               *
 *                                                                             *
 * Licensed under the Apache License, Version 2.0 (the "License");             *
 * you may not use this file except in compliance with the License.            *
 * You may obtain a copy of the License at                                     *
 *                                                                             *
 * http://www.apache.org/licenses/LICENSE-2.0                                  *
 *                                                                             *
 * Unless required by applicable law or agreed to in writing, software         *
 * distributed under the License is distributed on an "AS IS" BASIS,           *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.    *
 * See the License for the specific language governing permissions and         *
 * limitations under the License.                                              *
 *                                                                             *
 *                                                                             *
 *                                                                             *
 *******************************************************************************/
import EventEmitter from 'events'
import type TypedEventEmitter from 'typed-emitter'
import net from 'net'
import { CoreMsg, GetDeviceInfoReq } from './_generated/mdif/core/core'

// {a:A} | {b:B, c:C} => 'a' | 'b' | 'c'
type KeysOfUnion<T> = T extends T ? keyof T : never
type KeysOfProtobuf<PB> = Exclude<KeysOfUnion<PB>, '$case'>
// PickType<{a:A} | {b:B, c:C}, 'c'>  => C
type PickType<T, K extends KeysOfUnion<T>> = T extends { [k in K]?: unknown } ? T[K] : never
// {a:A} | {b:B} => {a:A, b:B}
type Merge<T> = {
    [k in KeysOfProtobuf<T>]: (arg: PickType<T, k>) => void
}

export type CoreMap = Merge<CoreMsg['msg']>

/// Event emitter mapping from protobuf socket to json
export const mdif = new EventEmitter() as TypedEventEmitter<CoreMap>

function decodeMdifMsg(msg: Buffer): void {
    const coreMsg = CoreMsg.decode(msg)
    if (!coreMsg.msg) {
        console.log('not a core message!')
        return
    }
    console.log('coreMsg', coreMsg)

    const event = coreMsg.msg.$case
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const data = coreMsg.msg[event]
    if (!event || !data) {
        console.log('invalid message:', { coreMsg, event, data })
        return
    }
    mdif.emit(event, data)
}

export async function mdifService(ip: string, port: number): Promise<void> {
    const socket = new net.Socket()
    socket.on('error', (err) => {
        console.log('Socket error', err)
        process.exit(2)
    })

    mdif.on('getDeviceInfoReq', (req) => {
        const coreMsg = CoreMsg.create()
        coreMsg.msg = { $case: 'getDeviceInfoReq', getDeviceInfoReq: GetDeviceInfoReq.create(req) }
        const b = Buffer.from(CoreMsg.encode(coreMsg).finish())
        const l = Buffer.alloc(4)
        l.writeUint32LE(b.length)
        socket.write(Buffer.concat([l, b]))
    })

    let buf: Buffer | null = null
    socket.on('data', (data) => {
        if (buf) {
            data = Buffer.concat([buf, data])
            buf = null
        }

        let length = data.length
        while (length >= 4) {
            const n = data.readUint32LE()
            if (length < n + 4) {
                break
            }
            length -= n + 4
            decodeMdifMsg(data.subarray(4, 4 + n))
            data = data.subarray(4 + n)
        }
        if (length) {
            buf = data.subarray(-length)
        }
    })

    return new Promise<void>((resolve) => {
        socket.connect(port, ip, () => {
            console.log('Connected to device')
            resolve()
        })
    })
}

File: src/android_demo/app/src/main/java/dk/mydefence/mdif_example/Mdif.java
/*******************************************************************************
 *                                                                             *
 *                                                 ,,                          *
 *                                                       ,,,,,                 *
 *                                                           ,,,,,             *
 *           ,,,,,,,,,,,,,,,,,,,,,,,,,,,,                        ,,,,          *
 *          ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,            ,,,,          ,,,,        *
 *          ,,,,,       ,,,,,      ,,,,,,                ,,,,        ,,,       *
 *          ,,,,,       ,,,,,      ,,,,,,                   ,,,        ,,,     *
 *          ,,,,,       ,,,,,      ,,,,,,       ,,,           ,,,        ,     *
 *          ,,,,,       ,,,,,      ,,,,,,           ,,,         ,,        ,    *
 *          ,,,,,       ,,,,,      ,,,,,,              ,,        ,,            *
 *          ,,,,,       ,,,,,      ,,,,,,                ,        ,            *
 *          ,,,,,       ,,,,,      ,,,,,,                 ,                    *
 *          ,,,,,       ,,,,,      ,,,,,,                                      *
 *          ,,,,,       ,,,,,      ,,,,,,                                      *
 *                                       ,,,,,,,,,,,,,,,,,,,,,,,,,,            *
 *                                       ,,,,,,,,,,,,,,,,,,,,,,,,,,,,          *
 *                                       ,,,,,                  ,,,,,,         *
 *                     ,                 ,,,,,                  ,,,,,,         *
 *             ,        ,,               ,,,,,                  ,,,,,,         *
 *    ,        ,,        ,,,             ,,,,,                  ,,,,,,         *
 *     ,        ,,,         ,,,          ,,,,,                  ,,,,,,         *
 *     ,,,       ,,,                     ,,,,,                  ,,,,,,         *
 *      ,,,        ,,,,                  ,,,,,                  ,,,,,,         *
 *        ,,,         ,,,,               ,,,,,                  ,,,,,,         *
 *         ,,,,,            ,,,,         ,,,,,,,,,,,,,,,,,,,,,,,,,,,,          *
 *            ,,,,                       ,,,,,,,,,,,,,,,,,,,,,,,,,,            *
 *               ,,,,,                                                         *
 *                    ,,,,,                                                    *
 *                                                                             *
 * Program/file : Mdif.java                                                    *
 *                                                                             *
 * Description  : Android implementation of MDIF                               *
 *              :                                                              *
 *                                                                             *
 * Copyright 2023 MyDefence A/S.                                               *
 *                                                                             *
 * Licensed under the Apache License, Version 2.0 (the "License");             *
 * you may not use this file except in compliance with the License.            *
 * You may obtain a copy of the License at                                     *
 *                                                                             *
 * http://www.apache.org/licenses/LICENSE-2.0                                  *
 *                                                                             *
 * Unless required by applicable law or agreed to in writing, software         *
 * distributed under the License is distributed on an "AS IS" BASIS,           *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.    *
 * See the License for the specific language governing permissions and         *
 * limitations under the License.                                              *
 *                                                                             *
 *                                                                             *
 *                                                                             *
 *******************************************************************************/
package dk.mydefence.mdif_example;

import android.content.Context;
import android.util.Log;

import dk.mydefence.mdif_example.hdlc.Hdlc;
import dk.mydefence.mdif_example.hdlc.IHdlc;

public class Mdif {
    private static final String TAG = "MD Mdif";
    private final Hdlc mHdlc;
    private final MdifCore mCore;
    private final IMdif mCallback;
    private final MdifRfs mRfs;

    protected enum MdifStatus {
        DECODE_SUCCESS,
        DECODE_ERR_NO_PAYLOAD,
        DECODE_ERR_NO_DECODER,
        DECODE_ERR_OTHER
    }

    public Mdif(Context applicationContext, IMdif mdifCallback, IMdifCore mdifCoreCallback, IMdifRfs rfsCallback) {
        mCallback = mdifCallback;
        mHdlc = new Hdlc(applicationContext, mHdlcCallback);
        mCore = new MdifCore(mHdlc, mdifCoreCallback);
        mRfs = new MdifRfs(mHdlc, rfsCallback);
    }

    private final IHdlc mHdlcCallback = new IHdlc() {
        @Override
        public void onHdlcFrame(byte[] frame) {
            if (mCore.decode(frame) != MdifStatus.DECODE_SUCCESS &&
                    mRfs.decode(frame) != MdifStatus.DECODE_SUCCESS) {
                Log.w(TAG, "Unable to decode received HDLC message");
            }
        }

        @Override
        public void onHdlcConnected() {
            Log.d(TAG, "HDLC connected - requesting device info");
            mCore.sendDeviceInfoReq();
            mCallback.onConnected();
        }

        @Override
        public void onHdlcDisconnected() {
            mCallback.onDisconnected();
        }

        @Override
        public void onHdlcConnecting() {
            mCallback.onConnecting();
        }

        @Override
        public void onHdlcFrameSent(byte[] frame) {
            Log.d(TAG, "Message sent through HDLC");
            mCallback.onMessageSent(frame);
        }
    };


    public void onResume() {
        mHdlc.onResume();
    }

    public void onPause() {
        mHdlc.onPause();
    }

    public void onDestroy() {
        mHdlc.onDestroy();
    }

}
